### PR TITLE
Remove Fizzler V3

### PR DESCRIPTION
**Pull request by inforithmics**
#### Reference Issue
https://github.com/svg-net/SVG/discussions/1073

#### What does this implement/fix? Explain your changes.
remove Fizzler implements a QuerySelectAll that uses the ExCSS Selector Dom

#### Any other comments?
This pull request is based on this pull request
https://github.com/svg-net/SVG/pull/1083

TODO: 
- [x] Pass All W3C Images with the same Selector results.
- [x] Add more Unit Tests
- [x] Add Benchmarks
- [x] Implement not Implemented Functions
- [x] Investigate AngleSharp Selector Integration (nuget or Code) 
- [x] Add more Selector Tests (Selector patterns from AngleSharp Unit Tests)  
- [x] Fix nth-child(2) (needs this pull request https://github.com/TylerBrinks/ExCSS/pull/162)
- [x] Implement :not (Is Implemented in Fizzler 1.3.0)
- [x] Implement nth-type and nth-last-type 
- [x] Investigate :target handling (I think in the context of svg rendering this isn't useful, so this is not implemented, because target only works with the selected url hash)
- [x] Investigate :root handling

A new Pull Request based on this.
https://github.com/svg-net/SVG/pull/1086

Main Idea:
The idea is to Implement an IExCssSelectorOps Interface similar interface to IElementOps<SvgElement> so that I can regression test to Fizzler Implementation. The ExCss Query tries to map the Selectors to the IExCssSelectorOps Interface, Because Fizzler has not implemented all Css Selectors certain Selectors are not implemented.

List of css selectors which should be supported.
https://stackoverflow.com/questions/42230563/what-is-the-compatibility-of-css3-pseudoclasses-with-svg-renderable-elements
All are handled (:target and :root in need to investigate)

### CODE REVIEW

**Comment by inforithmics** [comment]
I have enough Implemented for AngleSharp to run the Benchmark. here are the numbers (Maybe I need some more optimization)

|                        Method |       Mean |     Error |    StdDev |      Op/s | Rank |    Gen0 |   Gen1 | Allocated |
|------------------------------ |-----------:|----------:|----------:|----------:|-----:|--------:|-------:|----------:|
|      SelectorPerformanceExCss |   5.188 us | 0.0733 us | 0.0685 us | 192,739.3 |    1 |  2.6550 | 0.0153 |   21.7 KB |
|    SelectorPerformanceFizzler |  21.200 us | 0.2856 us | 0.2672 us |  47,169.1 |    2 |  3.7537 |      - |   30.9 KB |
| SelectorPerformanceAngleSharp | 264.828 us | 5.1335 us | 5.7059 us |   3,776.0 |    3 | 49.8047 | 2.9297 | 408.43 KB |

There is certainly room for improvements in the AngleSharp Implementation. But I don't know if I can find there Factor 50.

**Comment by inforithmics** [comment]
Implmented N th Child and N th Last Child with reflection, if the Properties become public in excss than this refelction Code can be changed to Field, Property Access.

https://github.com/TylerBrinks/ExCSS/pull/162

All Css Features that Fizzler supported are now supported. By the Excss Evalutor.

1. I test css selectors from AngeSharp so that the behave the same in Fizzler and excss.
2. Parse all svg documents and runt the css Selectors to test if they behave the same.

File: Source/Css/ExCssQuery.cs [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+                ChildSelector childSelector => ops.Child(),
+                ListSelector listSelector => GetFunc(listSelector, ops, inFunc), // TODO:,
+                NamespaceSelector namespaceSelector => throw new NotImplementedException(), // TODO:,
+                PageSelector pageSelector => throw new NotImplementedException(), // TODO:,
+                PseudoClassSelector pseudoClassSelector => GetFunc(pseudoClassSelector, ops, inFunc),
+                PseudoElementSelector pseudoElementSelector => throw new NotImplementedException(), // TODO:,
+                TypeSelector typeSelector => ops.Type(typeSelector.Name),
+                UnknownSelector unknownSelector => throw new NotImplementedException(), // TODO:,
+                IdSelector idSelector => ops.Id(idSelector.Id),
+                _ => throw new NotImplementedException(), // TODO:,
```
Comment (H1Gdev) [inline Source/Css/ExCssQuery.cs:263]: TODO exists.

File: Source/SvgDocument.cs [hunk 2]
```diff
@@ -421,6 +462,7 @@ private static T Create<T>(XmlReader reader) where T : SvgDocument, new()
                             {
                                 styles.Add(unknown);
                             }
+
```
Comment (H1Gdev) [inline Source/SvgDocument.cs:465]: remove this line.

Comment (inforithmics) [reply Source/SvgDocument.cs:465]: done

File: Source/Css/ExCssQuery.cs (hunk 1 above)
Comment (inforithmics) [reply Source/Css/ExCssQuery.cs]: Removed the todos and add one more case that is not implemented. For the not typed selectors I have to look into excss what kind of selectors this is.

File: Source/Css/ExCssQuery.cs [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+            {
+                AllSelector allSelector => ops.Universal(),
+                AttrAvailableSelector attrAvailableSelector => ops.AttributeExists(attrAvailableSelector.Attribute),
+                AttrBeginsSelector attrBeginsSelector => ops.AttributePrefixMatch(attrBeginsSelector.Attribute, attrBeginsSelector.Value),
+                AttrContainsSelector attrContainsSelector => ops.AttributeSubstring(attrContainsSelector.Attribute, attrContainsSelector.Value),
+                AttrEndsSelector attrEndsSelector => ops.AttributeSuffixMatch(attrEndsSelector.Attribute, attrEndsSelector.Value),
+                AttrHyphenSelector attrHyphenSelector => ops.AttributeDashMatch(attrHyphenSelector.Attribute, attrHyphenSelector.Value),
+                AttrListSelector attrListSelector => ops.AttributeIncludes(attrListSelector.Attribute, attrListSelector.Value),
+                AttrMatchSelector attrMatchSelector => ops.AttributeExact(attrMatchSelector.Attribute, attrMatchSelector.Value),
+                AttrNotMatchSelector attrNotMatchSelector => ops.AttributeNotMatch(attrNotMatchSelector.Attribute, attrNotMatchSelector.Value), // TODO:,
```
Comment (H1Gdev) [inline Source/Css/ExCssQuery.cs:244]: TODO exists.

**Comment by mrbean-bremen** [comment]
From my perspective this looks quite good, but I'm not really qualified to review this. @wieslawsoltes and @H1Gdev - please check if this can be merged.

File: Source/Css/ExCssQuery.cs (hunk 3 above)
Comment (inforithmics) [reply Source/Css/ExCssQuery.cs]: removed because it is now implemented

**Comment by H1Gdev** [comment]
@inforithmics 
What is AngleSharp used for ?

**Comment by inforithmics** [comment]
Currently nothing, I used it for the Benchmarks. But when I saw that it was not very Performant I didn't continue investigating it. So I remove the AngleSharp Code from this pull request.

**Comment by inforithmics** [comment]
So I reduced the changes and removed AngleSharp Code

**Comment by inforithmics** [comment]
I Implemented the :not Css Selector because in newest Fizzler 1.3.0 this was implemted (To keep feature parity).

File: Tests/Svg.UnitTests/Svg.UnitTests.csproj [hunk 4]
```diff
@@ -36,11 +36,12 @@
   </ItemGroup>
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.8.3" />
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.11.0" />
```
Comment (inforithmics) [inline Tests/Svg.UnitTests/Svg.UnitTests.csproj:39]: 16.8.3 has a security vulnerability so I upgraded to 16.11.0 to silence the warning.

File: Tests/Svg.UnitTests/Svg.UnitTests.csproj [hunk 5]
```diff
@@ -36,11 +36,12 @@
   </ItemGroup>
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.8.3" />
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.11.0" />
     <PackageReference Include="NUnit" Version="3.13.0" />
     <PackageReference Include="NUnit3TestAdapter" Version="3.17.0" />
     <PackageReference Include="Microsoft.NETFramework.ReferenceAssemblies" Version="1.0.0" PrivateAssets="All" />
     <PackageReference Include="Moq" Version="4.16.1" />
+    <PackageReference Include="Fizzler" Version="1.3.0" />
```
Comment (inforithmics) [inline Tests/Svg.UnitTests/Svg.UnitTests.csproj:44]: Upgraded to newest fizzler (for regression tests)

File: Tests/Svg.UnitTests/Css/SvgElementOps.cs [hunk 6]
```diff
@@ -177,7 +177,7 @@ private IEnumerable<SvgElement> ElementsAfterSelf(SvgElement self)
 
         public Selector<SvgElement> NthLastChild(int a, int b)
         {
-            throw new NotImplementedException();
```
Comment (inforithmics) [inline Tests/Svg.UnitTests/Css/SvgElementOps.cs:180]: nlast child was not implemented previously is now implemented.

File: Source/Css/ExCssQuery.cs [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+            {
+                pseudoFunc = ops.OnlyChild();
+            }
+            else
+            {
+                if (selector.Class.StartsWith(PseudoClassNames.Not))
+                {
+                    var sel = selector.Class.Substring(PseudoClassNames.Not.Length + 1, selector.Class.Length - 2 - PseudoClassNames.Not.Length);
+                    var parser = new StylesheetParser(true, true, tolerateInvalidValues: true);
+                    var styleSheet = parser.Parse(sel);
```
Comment (inforithmics) [inline Source/Css/ExCssQuery.cs:134]: excss does not fully parse the not pseudofunc so I have to parse it myself here.

File: README.md [hunk 8]
```diff
@@ -27,5 +27,4 @@ This project has dependencies on other open-source projects. These projects are
 
 |Project|Author|Sources|License|
 |--------|-----|---|---------|
-|Fizzler|Atif Aziz (@atifaziz)|[GitHub](https://github.com/atifaziz/Fizzler)|[LGPL](https://github.com/atifaziz/Fizzler/blob/master/COPYING.txt)|
```
Comment (inforithmics) [inline README.md:30]: is now removed from the nuget package as dependency is now only used as unit test dependency.

**Review by H1Gdev — APPROVED** [review]

**Pull request merged**